src/object/cli_obj.c

```diff
@@ -1530,49 +1470,494 @@ obj_shard_task_sched(struct obj_auxi_args *obj_auxi)
 		tse_task_complete(obj_auxi->obj_task, 0);
 }
 
-struct shard_reset_target {
-	unsigned int		 srt_is_punch:1;
-	unsigned int		 srt_rdg_count;
-	unsigned int		*srt_shards;
-};
+static struct shard_auxi_args *
+obj_embedded_shard_arg(struct obj_auxi_args *obj_auxi)
+{
+	switch (obj_auxi->opc) {
+	case DAOS_OBJ_RPC_UPDATE:
+	case DAOS_OBJ_RPC_FETCH:
+		return &obj_auxi->rw_args.auxi;
+	case DAOS_OBJ_RPC_PUNCH:
+	case DAOS_OBJ_RPC_PUNCH_DKEYS:
+	case DAOS_OBJ_RPC_PUNCH_AKEYS:
+		return &obj_auxi->p_args.pa_auxi;
```

Comment:
then why don't we just define the shard_auxi_args as a direct member of obj_auxi_args?

Reply:
because for example in shard_update_prep/shard_punch_prep it needs to set some other shard args that is not in shard_auxi_args.
actually I want to change current daos_obj_update_t and daos_obj_fetch_t to be one struct daos_obj_rw_t that would make some parameter handling be simpler, but can be done at following patch.